lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -62,23 +62,31 @@ def get_paginated_obj(self, query: Query, json: JSON, url: str, page_limit: int)
         """
         paginated_result = {}
         total_count = query.count()
-        if total_count > page_limit:
-            if "page_no" in json:
-                items = (
-                    query.order_by(Dataset.name)
-                    .limit(page_limit)
-                    .offset((json["page_no"] - 1) * page_limit)
-                    .all()
-                )
-                json["page_no"] += 1
-            else:
-                # This is the first page
-                items = query.limit(page_limit).all()
-                json["page_no"] = 1
-            paginated_result["next_url"] = url.rsplit("?", 1)[0] + "?" + urlencode(json)
-            paginated_result["total"] = total_count
+        page_limit = json.get(
+            "page_limit", int(self.config.server_config.get("page_limit"))
+        )
+        page_no = json.get("page_no", 1)
+        items = (
+            query.order_by(Dataset.name)
+            .limit(page_limit)
+            .offset((page_no - 1) * page_limit)
+            .all()
+        )
+        json["page_no"] = page_no + 1
+        parsed_url = urlparse(url)
+        if len(items) < page_limit:
+            next_url = ""
```

Comment:
Are you sure that you want the server-side of the request to work in "pages"?  It has a couple of downsides:
- we (apparently) have to have a server configuration option to set the default page size;
- we have to convert between "page numbers" and item offsets; and,
- it produces unpleasant results if the page size varies between requests.

If we tracked things by items/lines, the code might look more like this:
```
        offset = json.get("item_offset", 0)
        item_count = json.get("item_count", 0)
        q = query.order_by(Dataset.name).offset(offset)
        if item_count:
            q = q.limit(item_count)
        items = q.all()
        n = len(items)
        json["item_offset"] = offset + n
        parsed_url = urlparse(url)
        if n < item_count:
            next_url = ""
```
- With this approach, we don't impose a limit on the number of items returned (unless the request does); thus, we don't need a server configuration option for that value.  (Ultimately, we will probably want to have some sort of a limit on the maximum size of _any_ response, but it will probably be in _bytes_ rather than **rows**.)
- Instead of passing back a page number, we pass back a row number.  This means that there is no conversion on the server side between items and pages, and the server doesn't need to know how big a page is, at all (and, the offset doesn't fall victim to changes in the page size).

Reply:
"Page" is a pretty common term in this sort of model, so there's some risk in arbitrarily avoiding it. "Rows" sounds too DB-focused. We could do "items" or something like that, but it seems a bit artificial.

Reply:
Yeah, as Dave said these terminologies are very common in the community when the pagination is used. My reason behind using "pages", "next_url", and "total" is to sort of mimic the Elasticsearch response which many people are familiar with. Returning row numbers could expose some internals about our database which I am not sure is necessarily a bad thing though.

Reply:
> Returning row numbers could expose some internals about our database which I am not sure is necessarily a bad thing though.

I think that returning our actual DB row values *would* be a bad thing, partly because we should be able to reconstruct our DB at need without affecting external APIs.

I think that in this context, however, "row" would refer to an ordinal displacement within the set of datasets selected by the query, *not* absolute PostgreSQL `id` column values. That is, the query covers 1,000 datasets, hypothetically enumerated 1 to 1000; we're now returning the 21st through the 40th, for example.

Reply:
Ah! That's right. So do we want to use lines/items or should we continue using the pages and count?